File: iosApp/iosApp/voicevox/VoicevoxNativeImpl.swift
import ComposeApp
import voicevox_core
import AVFoundation
import UIKit

class VoicevoxNativeImpl: NSObject, VoicevoxNative {
    var libraryDir: String {
        get {
            FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first!.path
        }
    }

    var cacheDir: String {
        get {
            FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!.path
        }
    }
    private var isVoiceModelLoaded = false
    private var player: AVAudioPlayer?
    private var playWavFileContinuation: CheckedContinuation<Void, Never>?

    func speak(text: String, wavFilePath: OkioPath, openJtalkDicDirPath: OkioPath, vvmFilePath: OkioPath, styleId: Int32) async throws {
        do {
            let synthesizer = try makeSynthesizer(openJtalkDictDir: openJtalkDicDirPath.description())
            try loadVoiceModelIfNeeded(synthesizer: synthesizer, vvmPath: vvmFilePath.description())
            let voicevoxStyleId = VoicevoxStyleId(styleId)

            // 音声合成
            guard let audioQuery = synthesizer.audioQuery(text: text, styleId: voicevoxStyleId) else {
                throw MyError.runtimeError("Failed to create audioQuery: \(text)")
            }
            let wavData = try synthesizer.synthesis(audioQuery: audioQuery, styleId: voicevoxStyleId, options: voicevox_make_default_synthesis_options())
            let wavFileURL = URL(fileURLWithPath: wavFilePath.description())
            try wavData.write(to: wavFileURL)

            // 再生
            try await playWavFile(wavFilePath: wavFilePath)
        } catch {
            print(error)
        }
    }

    private func makeSynthesizer(openJtalkDictDir: String) throws -> Synthesizer {
        let onnxruntime = try Onnxruntime()
        let openJtalkDictDirURL = URL(fileURLWithPath: openJtalkDictDir)
        let openJtalk = try OpenJtalkRc(openJtalkDic: openJtalkDictDirURL)
        let initializeOptions: VoicevoxInitializeOptions = voicevox_make_default_initialize_options()
        return try Synthesizer(onnxruntime: onnxruntime, openJtalk: openJtalk, options: initializeOptions)
    }

    private func loadVoiceModelIfNeeded(synthesizer: Synthesizer, vvmPath: String) throws {
        if isVoiceModelLoaded { return }
        let voiceModelURL = URL(fileURLWithPath: vvmPath)
        let voiceModel = try VoiceModel(voiceModelURL: voiceModelURL)
        try synthesizer.loadVoiceModel(voiceModel: voiceModel)
    }

    private func playWavFile(wavFilePath: OkioPath) async throws -> Void {
        return await withCheckedContinuation { continuation in
            playWavFileContinuation = nil
            let fileURL = URL(fileURLWithPath: wavFilePath.description())
            do {
                let wavData = try Data(contentsOf: fileURL, options: [])
                // 再生
                if player?.isPlaying == true {
                    player?.stop()
                }
                player = try? AVAudioPlayer(data: wavData)
                player?.delegate = self
                player?.play()
                playWavFileContinuation = continuation
            } catch {
                print("Failed to read file")
                continuation.resume()
            }
        }
    }
}

extension VoicevoxNativeImpl: AVAudioPlayerDelegate {
    func audioPlayerDidFinishPlaying(_: AVAudioPlayer, successfully: Bool) {
        playWavFileContinuation?.resume()
        playWavFileContinuation = nil
    }
}

File: composeApp/src/androidMain/kotlin/app/example/voicevox/VoicevoxNativeImpl.kt
package app.example.voicevox

import android.media.MediaPlayer
import app.example.MainActivity
import co.touchlab.kermit.Logger
import jp.hiroshiba.voicevoxcore.AccelerationMode
import jp.hiroshiba.voicevoxcore.blocking.Onnxruntime
import jp.hiroshiba.voicevoxcore.blocking.OpenJtalk
import jp.hiroshiba.voicevoxcore.blocking.Synthesizer
import jp.hiroshiba.voicevoxcore.blocking.VoiceModelFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.Path
import java.io.File
import java.io.FileOutputStream
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class VoicevoxNativeImpl : VoicevoxNative {
    override val libraryDir: String
        get() = applicationContext.filesDir.toString()
    override val cacheDir: String
        get() = applicationContext.cacheDir.toString()
    private val applicationContext get() = MainActivity.INSTANCE.applicationContext
    private var mediaPlayer: MediaPlayer? = null
    private var isVoiceModelLoaded = false

    override suspend fun speak(text: String, wavFilePath: Path, openJtalkDicDirPath: Path, vvmFilePath: Path, styleId: Int) {
        withContext(Dispatchers.IO) {
            val synthesizer = makeSynthesizer(openJtalkDicDirPath)
            loadVoiceModelIfNeeded(synthesizer, vvmFilePath)

            // wavファイル作成
            val wavFile = File(wavFilePath.toString())

            // 音声合成
            val audioQuery = try {
                synthesizer.createAudioQuery(text, styleId)
            } catch (e: Exception) {
                Logger.e("Failed to createAudioQuery", e)
                return@withContext
            }
            val wavData = synthesizer.synthesis(audioQuery, styleId).perform()
            val fos = FileOutputStream(wavFile)
            fos.write(wavData)
            fos.close()

            // 再生
            playWavFile(wavFilePath)
        }
    }

    private fun makeSynthesizer(openJtalkDictDirPath: Path): Synthesizer {
        val onnxruntime = Onnxruntime.loadOnce().perform()
        val openJtalk = OpenJtalk(openJtalkDictDirPath.toString())
        val synthesizer = Synthesizer.builder(onnxruntime, openJtalk)
            .accelerationMode(
                AccelerationMode.AUTO
            )
            .build()
        return synthesizer
    }

    private fun loadVoiceModelIfNeeded(synthesizer: Synthesizer, vvmPath: Path) {
        if (isVoiceModelLoaded) return
        try {
            val voiceModel = VoiceModelFile(vvmPath.toString())
            synthesizer.loadVoiceModel(voiceModel)
        } catch (e: Exception) {
            Logger.e("Failed to loadVoiceModel", e)
        }
    }

    private suspend fun playWavFile(wavFilePath: Path) = suspendCoroutine { continuation ->
        mediaPlayer?.stop()
        mediaPlayer?.release()

        val mediaPlayer = MediaPlayer()
        this.mediaPlayer = mediaPlayer
        mediaPlayer.setOnCompletionListener {
            mediaPlayer.stop()
            mediaPlayer.release()
            this.mediaPlayer = null
            continuation.resume(Unit)
        }
        mediaPlayer.setDataSource(wavFilePath.toString())
        mediaPlayer.prepare()
        mediaPlayer.start()
    }
}
